File: Battleships/Assets/Scripts/Engine/Ship.cs
using System;
using System.Linq;
using System.Collections.Generic;

namespace Engine
{
    [Serializable]
    public class Ship
    {
        [Serializable]
        public class Block
        {
            public Coords pos;
            public int damage;

            public Block(Coords pos)
            {
                this.pos = pos;
                this.damage = 0;
            }

            public Block(UnityEngine.Vector2Int pos) : this((Coords)pos) { }
        }

        public List<Block> blocks;
        public int health;

        public Ship()
        {
            this.blocks = new List<Block>();
        }

        public bool IsSunk()
        {
            return this.blocks.All(block => block.damage > 95);
        }
    }
}

File: battleships_engine/src/ship.rs
use crate::coords::Coords;

#[derive(Serialize, Deserialize, Debug)]
pub struct Ship {
    pub blocks: Vec<Block>,
    pub health: i32,
}

impl Ship {
    pub fn new(health: i32) -> Self {
        Self {
            blocks: Vec::new(),
            health,
        }
    }

    pub fn add_block(&mut self, x: i32, y: i32) {
        self.blocks.push(Block::new(x, y));
    }

    pub fn is_sinked(&self) -> bool {
        self.blocks.iter().all(|block| block.damage > 95)
    }
}

#[derive(Default, Serialize, Deserialize, Debug)]
pub struct Block {
    pub pos: Coords,
    pub damage: i32,
}

impl Block {
    pub fn new(x: i32, y: i32) -> Self {
        Self {
            pos: Coords::new(x, y),
            damage: 0,
        }
    }
}

impl From<Coords> for Block {
    fn from(pos: Coords) -> Self {
        Self::new(pos.x, pos.y)
    }
}
